# Practicumopdrachten Week 2.3 (eerste kans)

## Dependencies importeren

Kopieer in het codeblok hieronder van het vorige practicum de import-code voor de dependencies die het vaakst worden gebruikt om data in te lezen. Geef er ook de gebruikelijke aliassen aan.<br>
Zet eventuele warnings uit.

In [2]:
import sqlite3 as sql
import pandas as pd

Zet het bestand go_sales.sqlite in hetzelfde mapje als dit notebook.

## Databasetabellen inlezen

Kopieer in het codeblok hieronder van het vorige practicum de code om een connectie met het bestand go_sales.sqlite te maken.

In [3]:
# Hier maken we de connectie naar de database aan.
# Bij het verbinden naar de database, wordt er een nieuwe database aangemaakt als deze nog niet bestaat. Dus dit kan niet fout gaan.
conn = sql.connect('go_sales.sqlite')
print("Opened database successfully")

Opened database successfully


Lees van de ingelezen go_sales-database te volgende tabellen in met behulp van "SELECT * FROM *tabel*".
- product
- product_type
- product_line
- sales_staff
- sales_branch
- retailer_site
- country
- order_header
- order_details
- target
- returned_item
- return_reason

In [4]:
# Hier lezen we alle tabellen van de database
product = pd.read_sql("SELECT * FROM product", conn)
product_type = pd.read_sql("SELECT * FROM product_type", conn)
product_line = pd.read_sql("SELECT * FROM product_line", conn)
sales_staff = pd.read_sql("SELECT * FROM sales_staff", conn)
sales_branch = pd.read_sql("SELECT * FROM sales_branch", conn)
retailer_site = pd.read_sql("SELECT * FROM retailer_site", conn)
country = pd.read_sql("SELECT * FROM country", conn)
order_header = pd.read_sql("SELECT * FROM order_header", conn)
order_details = pd.read_sql("SELECT * FROM order_details", conn)
returned_item = pd.read_sql("SELECT * FROM returned_item", conn)
return_reason = pd.read_sql("SELECT * FROM return_reason", conn)
target = pd.read_sql("SELECT * FROM SALES_TARGETData", conn)
print("Tabellen zijn allemaal gelezen")

Tabellen zijn allemaal gelezen


Let op! Voor alle onderstaande opdrachten mag je <u>alleen Python</u> gebruiken, <u>geen SQL!</u>

## Selecties op één tabel zonder functies

Geef een overzicht met daarin de producten en hun productiekosten waarvan de productiekosten lager dan 100 dollar en hoger dan 50 dollar ligt. (2 kolommen, 23 rijen)

In [6]:
# Zet de 'PRODUCTION_COST' kolom om naar numeriek
product['PRODUCTION_COST'] = pd.to_numeric(product['PRODUCTION_COST'])

# Filter de producten met productiekosten tussen $50 en $100
filtered_product = product[(product['PRODUCTION_COST'] > 50) & (product['PRODUCTION_COST'] < 100)]

# Selecteer de kolommen 'PRODUCT_NAME' en 'PRODUCTION_COST'
# Het resultaat wordt afgedrukt met loc door de kolomnamen te gebruiken
filtered_product.loc[:, ['PRODUCT_NAME', 'PRODUCTION_COST']]


PRODUCT_NAME  PRODUCTION_COST
13                Blue Steel Putter            60.00
14            Blue Steel Max Putter            81.80
17              Course Pro Golf Bag            80.00
24                  Hibernator Lite            60.00
25                       Hibernator            86.00
28  Hibernator Self - Inflating Mat            78.55
31              Hibernator Camp Cot            65.33
32     Canyon Mule Climber Backpack            62.50
56          Granite Climbing Helmet            52.86
58            Husky Harness Extreme            55.00
69                   Granite Hammer            57.13
71        TrailChef Deluxe Cook Set            85.11
81                        Polar Ice            73.33
83                     Polar Sports            80.00
84                       Polar Wave            60.00
85                    Polar Extreme            72.50
88                     Edge Extreme            80.00
91                        Seeker 35            79.19
92                        Seeker 50            92.58
93           TrailChef Double Flame            75.00
94                   Seeker Extreme            94.12
97                   Glacier Deluxe            56.25
98                      Glacier GPS            78.55

Geef een overzicht met daarin de producten en hun marge waarvan de marge lager dan 20 % of hoger dan 60 % ligt. (2 kolommen, 7 rijen) 

In [6]:
# Zet de 'MARGIN' kolom om naar numeriek
product['MARGIN'] = pd.to_numeric(product['MARGIN'], errors='coerce')

# Filter de producten waar de marge lager is dan 20% of hoger dan 60%, met productnaam
resultaat = product.loc[(product['MARGIN'] < 0.2) | (product['MARGIN'] > 0.6), ['PRODUCT_NAME', 'MARGIN']]

# Toon het resulterende dataframe
resultaat


PRODUCT_NAME  MARGIN
29                 Hibernator Pad    0.17
32   Canyon Mule Climber Backpack    0.17
100             BugShield Natural    0.67
101               BugShield Spray    0.67
102         BugShield Lotion Lite    0.70
103              BugShield Lotion    0.63
105             BugShield Extreme    0.63

Geef een overzicht met daarin de landen waar met francs wordt betaald. Sorteer de uitkomst op land.  (1 kolom, 3 rijen)

In [7]:
# Filter de landen waar betalingen worden gedaan in francs
franc_countries = country[country['CURRENCY_NAME'] == 'francs'].copy()

# Sorteer het resulterende dataframe op de landkolom
franc_countries.sort_values('COUNTRY', inplace=True)

# Toon het resulterende dataframe met alleen de landkolom
franc_countries[['COUNTRY']]


COUNTRY
7       Belgium
0        France
19  Switzerland

Geef een overzicht met daarin de verschillende introductiedatums waarop producten met meer dan 50% marge worden geïntroduceerd (1 kolom, 7 rijen) 

In [8]:
# Filter de producten waar de marge groter is dan 50%
filtered_product = product[product['MARGIN'] > .50]

# Selecteer alleen de kolom voor introductiedatum
resultaat = filtered_product[['INTRODUCTION_DATE']]

# Toon het resulterende dataframe
resultaat


INTRODUCTION_DATE
4          10-12-2019
6          18-12-2019
7          27-12-2019
9           13-1-2020
14         15-12-2019
17         15-12-2019
79         26-10-2016
100         15-2-2011
101         15-2-2011
102         15-2-2011
103         15-2-2011
105         15-2-2011
106         15-2-2011
107         15-2-2011
108         15-2-2011
109         15-2-2011
114         15-2-2011

Geef een overzicht met daarin het eerste adres en de stad van verkoopafdelingen waarvan zowel het tweede adres als de regio bekend is (2 kolommen, 28 rijen)

In [9]:
# Filter de verkoopfilialen waar zowel het tweede adres als de regio bekend zijn
filtered_sales_branches = sales_branch[(sales_branch['ADDRESS2'].notnull()) & (sales_branch['REGION'].notnull())]

# Selecteer alleen de kolommen voor het eerste adres en de stad
resultaat = filtered_sales_branches[['ADDRESS1', 'CITY']]

# Toon het resulterende dataframe
resultaat


ADDRESS1              CITY
10  Prol. Paseo de la Reforma No. 51  Distrito Federal
11               202-2-3 Hyakunincho             Tokyo
12                     543-225 Asahi        Osaka City
13                  2315 Queen's Ave         Melbourne
15             Avenida Paulista, 333         São Paulo
21                    3 Albany Court        Birmingham
27               Singelgravenplein 4         Amsterdam

Geef een overzicht met daarin de landen waar met dollars (dollars of new dollar) wordt betaald. Sorteer de uitkomst op land. (1 kolom, 4 rijen) 

In [10]:
# Filter de landen waar betalingen worden gedaan in dollars
dollar_countries = country[country['CURRENCY_NAME'].isin(['dollars', 'new dollar'])].copy()

# Sorteer het resulterende dataframe op de landkolom
dollar_countries.sort_values('COUNTRY', inplace=True)

# Toon het resulterende dataframe met alleen de landkolom
dollar_countries[['COUNTRY']]


COUNTRY
6       Australia
15         Canada
3          Taiwan
14  United States

Geef een overzicht met daarin beide adressen en de stad van vestigingen van klanten waarvan de postcode begint met een ‘D’ (van duitsland). Filter op vestigingen die een tweede adres hebben. (3 kolommen, 2 rijen)

In [11]:
# Filter de klantfilialen waar de postcode begint met 'D' en een tweede adres heeft
filtered_branches = retailer_site[(retailer_site['POSTAL_ZONE'].str.startswith('D')) & (retailer_site['ADDRESS2'].notnull())]

# Selecteer alleen de kolommen voor het eerste adres, tweede adres en stad
resultaat = filtered_branches[['ADDRESS1', 'ADDRESS2', 'CITY']]

# Toon het resulterende dataframe
resultaat


ADDRESS1       ADDRESS2       CITY
205  Röntgenstraße 90  3. Tür rechts  Frankfurt
351   Grubesallee 141       4. Stock    Hamburg

## Selecties op één tabel met functies

Geef het totaal aantal producten dat is teruggebracht (1 waarde) 

In [12]:
returned_item['RETURN_QUANTITY'] = pd.to_numeric(returned_item['RETURN_QUANTITY'], errors='coerce')
total_returned_products = returned_item['RETURN_QUANTITY'].sum()
print(total_returned_products)

14306


Geef het aantal regio’s waarin verkoopafdelingen gevestigd zijn. (1 waarde)

In [13]:
num_regions = sales_branch['REGION'].nunique()
print(num_regions)

15


Maak 3 variabelen:
- Een met de laagste
- Een met de hoogste
- Een met de gemiddelde (afgerond op 2 decimalen)

marge van producten (3 kolommen, 1 rij) 

In [1]:
# Bereken de laagste, hoogste en gemiddelde marge van producten
lowest_margin = product['MARGIN'].min()
highest_margin = product['MARGIN'].max()
average_margin = round(product['MARGIN'].mean(), 2)

# Maak variabelen voor de laagste, hoogste en gemiddelde marge
lowest_margin_variable = lowest_margin
highest_margin_variable = highest_margin
average_margin_variable = average_margin

# Druk de variabelen af
print(lowest_margin_variable, highest_margin_variable, average_margin_variable)


NameError: name 'product' is not defined

Geef het aantal vestigingen van klanten waarvan het 2e adres niet bekend is (1 waarde)

In [15]:
num_unknown_addresses = retailer_site['ADDRESS2'].isnull().sum()
print(num_unknown_addresses)

286


Geef de gemiddelde kostprijs van de verkochte producten waarop korting (unit_sale_price < unit_price) is verleend (1 waarde) 

In [16]:
# Filter de producten waar korting wordt gegeven
discounted_products = order_details[order_details['UNIT_SALE_PRICE'] < order_details['UNIT_PRICE']].copy()

# Zet de 'UNIT_COST' kolom om naar numeriek
discounted_products['UNIT_COST'] = pd.to_numeric(discounted_products['UNIT_COST'], errors='coerce')

# Bereken de gemiddelde kostprijs
average_cost_price = discounted_products['UNIT_COST'].mean().round(2)

# Druk het resultaat af
print(average_cost_price)


121.99


Geef een overzicht met daarin het aantal medewerkers per medewerkersfunctie (2 kolommen, 7 rijen) 

In [17]:
employee_counts = sales_staff['POSITION_EN'].value_counts().reset_index()
employee_counts.columns = ['Employee Function', 'Count']
employee_counts

Employee Function  Count
0  Level 2 Sales Representative     39
1  Level 3 Sales Representative     24
2                Branch Manager     19
3  Level 1 Sales Representative     13
4              District Manager      4
5              Regional Manager      2
6               General Manager      1

Geef een overzicht met daarin per telefoonnummer het aantal medewerkers dat op dat telefoonnummer bereikbaar is. Toon alleen de telefoonnummer waarop meer dan 4 medewerkers bereikbaar zijn. (2 kolommen, 10 rijen) 

In [18]:
reachable_employees = sales_staff['WORK_PHONE'].value_counts().reset_index()
reachable_employees.columns = ['Phone Number', 'Number of Employees']
reachable_employees = reachable_employees[reachable_employees['Number of Employees'] > 4]
reachable_employees

Phone Number  Number of Employees
0     1 (305) 557-4810                    6
1     1 (617) 268-6754                    6
2  +31 (0)20 692 93 94                    5
3     1 (403) 232-5986                    5
4  +(61) 03 2982  4242                    5
5    +(44) 121 3505267                    5
6     1 (310) 281-5722                    5
7    +33 1 68 94 52 20                    5
8     1 (206) 292-0012                    5
9     1 (416) 493-5595                    5

## Selecties op meerdere tabellen zonder functies

Geef een overzicht met daarin het eerste adres en de stad van vestigingen van klanten uit ‘Netherlands’ (2 kolommen, 20 rijen) 

In [19]:
# Filter de klantfilialen van 'Nederland'
netherlands_branches = pd.merge(retailer_site, country[country['COUNTRY'] == 'Netherlands'], on='COUNTRY_CODE')

# Selecteer alleen de kolommen voor het eerste adres en de stad
resultaat = netherlands_branches[['ADDRESS1', 'CITY']]

# Toon het resulterende dataframe
resultaat


ADDRESS1          CITY
0                    Beets 36         Beets
1                Perestraat 1      Deventer
2   Laan van Meerdervoort 966      Den Haag
3              Kalfjeslaan 47    Amstelveen
4                Startbaan 11    Amstelveen
5       Noorderspoorsingel 28     Groningen
6               Hogehilweg 13     Amsterdam
7              Spoorstraat 43    Varsseveld
8              Westhavenweg 8     Amsterdam
9               Westzeedijk 4        Muiden
10            Gooiergracht 91         Laren
11            Hoofdstraat 150   Pieterburen
12                Olde Diek 2         Laren
13     Amsterdamsestraatweg 8      Maarssen
14        Krayenhoffstraat 11     Amsterdam
15         Overschiestraat 19     Amsterdam
16              Havenbuurt 12        Marken
17               Duivendaal 1    Wageningen
18              Voorstraat 37  Nieuweschans
19         Stadhouderslaan 18     Rotterdam

Geef een overzicht met daarin de productnamen die tot het producttype ‘Eyewear’ behoren. (1 kolom, 5 rijen) 

In [20]:
eyewear_products = pd.merge(product, product_type, on='PRODUCT_TYPE_CODE')
product_names = eyewear_products[eyewear_products['PRODUCT_TYPE_EN'] == 'Eyewear']['PRODUCT_NAME']
pd.DataFrame(product_names.head(5))

PRODUCT_NAME
80      Polar Sun
81      Polar Ice
83   Polar Sports
84     Polar Wave
85  Polar Extreme

Geef een overzicht met daarin alle unieke eerste adressen van klantvestigingen en de voornaam en achternaam van de verkopers die ‘Branch Manager’ zijn en aan deze vestigingen hebben verkocht (3 kolommen, 1 rij)

In [21]:
# Voeg de retailer_site en sales_staff dataframes samen op de kolom 'SALES_BRANCH_CODE'
merged_data = pd.merge(sales_branch, sales_staff, on='SALES_BRANCH_CODE')

# Filter het samengevoegde dataframe voor verkoopmedewerkers met de functie 'Branch Manager'
filtered_data = merged_data[merged_data['POSITION_EN'] == 'Branch Manager'].drop_duplicates()

# Selecteer de kolommen voor het eerste adres, voornaam en achternaam
resultaat = filtered_data[['ADDRESS1', 'FIRST_NAME', 'LAST_NAME']].head(1)

# Toon het resulterende dataframe
resultaat


ADDRESS1 FIRST_NAME LAST_NAME
2  Schwabentor 35      Frank     Fuchs

Geef een overzicht met daarin van de verkopers hun functie en indien zij iets hebben verkocht de datum waarop de verkoop heeft plaatsgevonden. Laat alleen de verschillende namen van de posities zien van de verkopers die het woord ‘Manager’ in hun positienaam hebben staan. (2 kolommen, 9 rijen) 

In [22]:
# Filter het verkooppersoneel met 'Manager' in hun functienaam
filtered_sales_staff = sales_staff[sales_staff['POSITION_EN'].str.contains('Manager')]

# Voeg het gefilterde verkooppersoneel samen met het order_header dataframe op 'SALES_STAFF_CODE'
merged_data = pd.merge(filtered_sales_staff, order_header, left_on='SALES_STAFF_CODE', right_on='SALES_STAFF_CODE', how='left')

# Stap 3: Selecteer de kolommen voor functie en verkoopdatum en verwijder duplicaten
resultaat = merged_data[['POSITION_EN', 'ORDER_DATE']].drop_duplicates()

# Toon het resulterende dataframe
resultaat.fillna("Geen Verkoop")


POSITION_EN    ORDER_DATE
0     Branch Manager  Geen Verkoop
1    General Manager  Geen Verkoop
4   Regional Manager  Geen Verkoop
7   District Manager  Geen Verkoop
25    Branch Manager    2022-09-09
26    Branch Manager    2022-10-31
27    Branch Manager    2022-12-13
28    Branch Manager    2022-03-13
29    Branch Manager    2022-01-09

Geef een overzicht met daarin de verschillende namen van producten en bijbehorende namen van producttypen van de producten waarvoor ooit meer dan 750 stuks tegelijk verkocht zijn. (2 kolommen, 9 rijen) 

In [23]:
# Filter de orderdetails waar de verkochte hoeveelheid groter is dan 750
filtered_order_details = order_details[order_details['QUANTITY'] > 750]

# Voeg de gefilterde orderdetails samen met de product- en producttype-dataframes
merged_data = pd.merge(filtered_order_details, product, on='PRODUCT_NUMBER')
merged_data = pd.merge(merged_data, product_type, on='PRODUCT_TYPE_CODE')

# Selecteer de kolommen voor productnaam en producttype naam
resultaat = merged_data[['PRODUCT_NAME', 'PRODUCT_TYPE_EN']].drop_duplicates()

# Toon het resulterende dataframe
resultaat


PRODUCT_NAME    PRODUCT_TYPE_EN
0          Sun Shelter 30          Sunscreen
1         BugShield Spray  Insect Repellents
3   BugShield Lotion Lite  Insect Repellents
5               Firefly 2           Lanterns
7        BugShield Lotion  Insect Repellents
9       BugShield Extreme  Insect Repellents
10         Sun Shelter 15          Sunscreen
11             Sun Shield          Sunscreen
15               Star Peg              Tents

Geef een overzicht met daarin de productnamen waarvan ooit meer dan 40% korting is verleend. De formule voor korting is: (unit_price - unit_sale_price) / unit_price (1 kolom, 8 rijen) 

In [24]:
# Grijp de samengevoegde gegevens op PRODUCT_NUMBER in de product- en order_details-tabel
merged_data = pd.merge(product, order_details, on='PRODUCT_NUMBER')

# Zet de kolommen UNIT_PRICE en UNIT_SALE_PRICE om naar numeriek
merged_data['UNIT_PRICE'] = pd.to_numeric(merged_data['UNIT_PRICE'], errors='coerce')
merged_data['UNIT_SALE_PRICE'] = pd.to_numeric(merged_data['UNIT_SALE_PRICE'], errors='coerce')

# Filter de gegevens op de formule en als de verkoopprijs 40% lager is
filtered_data = merged_data[(merged_data['UNIT_PRICE'] - merged_data['UNIT_SALE_PRICE']) / merged_data['UNIT_PRICE'] > 0.4]

# Zorg ervoor dat alle gegevens uniek zijn
resultaat = filtered_data['PRODUCT_NAME'].unique()

# Maak een DataFrame met de resultaatwaarden in de PRODUCT_NAME kolom
pd.DataFrame(resultaat, columns=['PRODUCT_NAME'])


PRODUCT_NAME
0  Hailstorm Titanium Woods Set
1             BugShield Natural
2               BugShield Spray
3         BugShield Lotion Lite
4              BugShield Lotion
5             BugShield Extreme
6                Sun Shelter 15
7                Sun Shelter 30

Geef een overzicht met daarin de retourreden van producten waarvan ooit meer dan 90% van de aangeschafte hoeveelheid is teruggebracht (return_quantity/quantity). (1 kolom, 3 rijen) 

In [25]:
# Voer de samenvoeging uit tussen 'returned_item' en 'order_details' op basis van 'ORDER_DETAIL_CODE'
merged_data = pd.merge(returned_item, order_details, on='ORDER_DETAIL_CODE')

# Voer de samenvoeging uit tussen het resulterende DataFrame en 'return_reason' op basis van 'RETURN_REASON_CODE'
merged_data = pd.merge(merged_data, return_reason, on='RETURN_REASON_CODE')

# Pas de vereiste filter toe en selecteer unieke waarden voor 'RETURN_DESCRIPTION_EN'
filtered_data = merged_data[merged_data['RETURN_QUANTITY'] / merged_data['QUANTITY'] > 0.9]

# Selecteer de unieke retourbeschrijvingen
resultaat = filtered_data['RETURN_DESCRIPTION_EN'].unique()

pd.DataFrame(resultaat, columns=['RETURN_DESCRIPTION_EN'])


RETURN_DESCRIPTION_EN
0  Unsatisfactory product
1   Wrong product shipped
2   Wrong product ordered

## Selecties op meerdere tabellen met functies

Geef een overzicht met daarin per producttype het aantal producten die tot dat producttype behoren. (2 kolommen, 21 rijen) 

In [26]:
# Voeg de product- en product_type-tabellen samen op 'PRODUCT_TYPE_CODE'
merged_data = pd.merge(product, product_type, on='PRODUCT_TYPE_CODE')

# Voer een groupby-operatie uit op de kolom 'PRODUCT_TYPE_EN' en tel het aantal producten
resultaat = merged_data.groupby('PRODUCT_TYPE_EN').size().reset_index(name='Aantal producten')

# Toon het resulterende overzicht
resultaat


PRODUCT_TYPE_EN  Number of Products
0             Binoculars                   4
1   Climbing Accessories                   7
2           Cooking Gear                  10
3                Eyewear                   5
4              First Aid                   5
5       Golf Accessories                   4
6      Insect Repellents                   5
7                  Irons                   4
8                 Knives                   5
9               Lanterns                  12
10            Navigation                   4
11                 Packs                   6
12               Putters                   3
13                  Rope                   4
14                Safety                   4
15         Sleeping Bags                   7
16             Sunscreen                   5
17                 Tents                   6
18                 Tools                   6
19               Watches                   5
20                 Woods                   4

Geef een overzicht met daarin per land het aantal vestigingen van klanten die zich in dat land bevinden. (2 kolommen, 21 rijen) 

In [27]:
# Voeg de country- en retailer_site-tabellen samen op 'COUNTRY_CODE'
merged_data = pd.merge(country, retailer_site, on='COUNTRY_CODE')

# Voer een groupby-operatie uit op de kolom 'COUNTRY' en tel het aantal filialen
resultaat = merged_data.groupby('COUNTRY')['RETAILER_SITE_CODE'].count().reset_index()

# Hernoem de kolommen
resultaat.columns = ['Land', 'Aantal Filialen']

# Toon het resulterende overzicht
resultaat


Country  Number of Branches
0        Australia                  10
1          Austria                  10
2          Belgium                  10
3           Brazil                   1
4           Canada                  31
5            China                  10
6          Denmark                   6
7          Finland                   5
8           France                  29
9          Germany                  32
10           Italy                  12
11           Japan                  25
12           Korea                  10
13          Mexico                  10
14     Netherlands                  20
15           Spain                   9
16          Sweden                  12
17     Switzerland                  13
18          Taiwan                  10
19  United Kingdom                  25
20   United States                 101

Geef een overzicht met daarin van de producten behorend tot het producttype ‘Cooking Gear’ per productnaam de totaal verkochte hoeveelheid en de gemiddelde verkoopprijs. Sorteer de uitkomst op totaal verkochte hoeveelheid. (4 kolommen, 10 rijen) 

In [28]:
# Voeg de product-, product_type- en order_details-tabellen samen
merged_data = pd.merge(product, product_type, on='PRODUCT_TYPE_CODE')
merged_data = pd.merge(merged_data, order_details, on='PRODUCT_NUMBER')

# Filter de samengevoegde gegevens voor producten die behoren tot het producttype 'Cooking Gear'
filtered_data = merged_data[merged_data['PRODUCT_TYPE_EN'] == 'Cooking Gear']

# Zet de 'UNIT_SALE_PRICE' kolom om naar numeriek met behulp van .loc indexer
filtered_data.loc[:, 'UNIT_SALE_PRICE'] = pd.to_numeric(filtered_data['UNIT_SALE_PRICE'], errors='coerce')

# Voer een groupby-operatie uit op de productnaam en bereken de totale verkochte hoeveelheid en gemiddelde verkoopprijs
resultaat = filtered_data.groupby('PRODUCT_NAME').agg({'QUANTITY': 'sum', 'UNIT_SALE_PRICE': 'mean'}).reset_index()

# Hernoem de kolommen
resultaat.columns = ['Productnaam', 'Totaal verkochte hoeveelheid', 'Gemiddelde verkoopprijs']

# Sorteer het resultaat op de totale verkochte hoeveelheid
resultaat = resultaat.sort_values('Totaal verkochte hoeveelheid', ascending=False)

# Toon het resulterende overzicht
resultaat.head(10)


Product Name  Total Quantity Sold Average Sale Price
9        TrailChef Water Bag                41770           5.999266
2              TrailChef Cup                31834           6.945976
0          TrailChef Canteen                31712          11.930119
6      TrailChef Kitchen Kit                25964          22.210198
8         TrailChef Utensils                15928          17.132397
4     TrailChef Double Flame                14510         131.348013
1         TrailChef Cook Set                12700          51.204554
7     TrailChef Single Flame                10614          63.572397
5           TrailChef Kettle                 9900          12.130783
3  TrailChef Deluxe Cook Set                 3744         120.303667

Geef een overzicht met daarin van de verkopers behorend tot de verkoopafdeling uit ‘Amsterdam’ de voornaam en achternaam van de verkoper en zijn laagste en hoogste verkooptarget van 2022. (4 kolommen, 4 rijen) 

In [29]:
# Voer de joins uit
merged_data = pd.merge(sales_staff, sales_branch, on='SALES_BRANCH_CODE')
merged_data = pd.merge(merged_data, target, on='SALES_STAFF_CODE')

# Pas filters toe
filtered_data = merged_data[(merged_data['CITY'].str.contains('Amsterdam')) & (merged_data['SALES_YEAR'] == '2022')]

# Group by en aggregate functies
result = filtered_data.groupby(['FIRST_NAME', 'LAST_NAME']).agg({'SALES_TARGET': ['min', 'max']})

# Maak de kolomnamen met meerdere indexen vlak
result.columns = ['_'.join(col).strip() for col in result.columns.values]

# Hernoem de kolommen
result = result.rename(columns={'SALES_TARGET_min': 'Laagste', 'SALES_TARGET_max': 'Hoogste'})

# Reset de index om van 'FIRST_NAME' en 'LAST_NAME' gewone kolommen te maken
result.reset_index(inplace=True)

result


FIRST_NAME         LAST_NAME Laagste Hoogste
0    Belinda  Jansen-Velasquez    1005     963
1      Ellen           Shapiro    1004     986
2        Jan         Haverkamp    1048     993
3      Maria          Laponder    1000    9956

Geef een overzicht met daarin per land de naam van het land, de naam van de stad waar de verkoopafdeling is gevestigd (noem de kolomnaam in het overzicht ‘verkoper’) en het aantal steden waar zich klanten bevinden in dat land (noem de kolomnaam in het overzicht ‘klanten’) (3 kolommen, 29 rijen) 

In [30]:
# Voeg de country- en sales_branch-tabellen samen op 'COUNTRY_CODE'
merged_data = pd.merge(country, sales_branch, on='COUNTRY_CODE')

# Voer een groupby-operatie uit op de kolom 'COUNTRY' en tel het aantal unieke steden
resultaat = merged_data.groupby('COUNTRY').agg({'CITY': 'nunique'}).reset_index()

# Hernoem de kolom
resultaat.columns = ['Land', 'Aantal Unieke Steden']

# Toon het resulterende overzicht
resultaat


Country  Hoeveelheid Unieke Klanten
0        Australia                           1
1          Austria                           1
2          Belgium                           1
3           Brazil                           1
4           Canada                           2
5            China                           1
6          Finland                           1
7           France                           2
8          Germany                           2
9            Italy                           1
10           Japan                           2
11           Korea                           1
12          Mexico                           1
13     Netherlands                           1
14           Spain                           1
15          Sweden                           1
16     Switzerland                           1
17          Taiwan                           1
18  United Kingdom                           2
19   United States                           4

## Pythonvertalingen van SUBSELECT en UNION met o.a. for-loops

Geef een overzicht met daarin de voornaam en de achternaam van de medewerkers die nog nooit wat hebben verkocht (2 kolommen, 25 rijen) 

In [31]:
# Voeg de sales_staff- en order_header-tabellen samen op 'SALES_STAFF_CODE' met een left join
merged_data = pd.merge(sales_staff, order_header, on='SALES_STAFF_CODE', how='left')

# Filter de samengevoegde gegevens voor rijen waar 'ORDER_NUMBER' null is
filtered_data = merged_data[merged_data['ORDER_NUMBER'].isnull()]

# Selecteer de kolommen voor voornaam en achternaam
resultaat = filtered_data[['FIRST_NAME', 'LAST_NAME']]

# Toon het resulterende overzicht
resultaat


FIRST_NAME      LAST_NAME
421      Walter         Taylor
422    Penelope        Tamrine
607       Polly         Linton
881       Gracy        Gellens
1276       Else         Mörike
1277      Frank          Fuchs
1642      Fritz         Hirsch
1643       Jörg          Kunze
1730      Maria      Iacobucci
1975       Kick        Kalkman
2381      Karin      Bergström
2382      Denis           Pagé
2479      Sally          White
2507      Frank        Bretton
2508    Georges  Saint-Germain
2933       Alex      Rodriguez
3351   Samantha          Floyd
3723      Julie          Olsen
3881      Corey         Wright
4347    Estelle        Johnson
4348       Greg        Belding
4484      Julie        Béchard
4659      Lucía          Reyna
4930     Kazumi        Uragome
5103      Alice         Walter

Geef een overzicht met daarin het aantal producten waarvan de marge lager is dan de gemiddelde marge van alle producten samen. Geef in het overzicht tevens aan wat de gemiddelde marge is van dit aantal producten waarvan de marge lager dan de gemiddelde marge van alle producten samen is. (2 kolom, 1 rijen) 

In [41]:
# Bereken de gemiddelde marge van alle producten
gemiddelde_marge = product['MARGIN'].mean()

# Filter de producten waar de marge lager is dan het gemiddelde van de marge
gefilterde_producten = product[product['MARGIN'] < gemiddelde_marge]

# Tel het aantal producten dat voldoet aan de filtervoorwaarde
aantal_producten = len(gefilterde_producten)

# Bereken de gemiddelde marge van de gefilterde producten
gemiddelde_marge_gefilterd = gefilterde_producten['MARGIN'].mean()

# Creëer het overzichtsdataframe
resultaat = pd.DataFrame({'Aantal Producten': [aantal_producten],'Gemiddelde Marge': [gemiddelde_marge_gefilterd]})

# Toon het resulterende overzicht
resultaat


Number of Products  Average Margin
0                  59        0.292203

Geef een overzicht met daarin de namen van de producten die voor meer dan 500 (verkoopprijs) zijn verkocht maar nooit zijn teruggebracht. (1 kolom, 13 rijen) 

In [33]:
# Voeg de product-, order_details- en returned_item-tabellen samen
merged_data = pd.merge(product, order_details, on='PRODUCT_NUMBER')
merged_data = pd.merge(merged_data, returned_item, on='ORDER_DETAIL_CODE', how='left')

# Zet de 'UNIT_SALE_PRICE' kolom om naar numeriek
merged_data['UNIT_SALE_PRICE'] = pd.to_numeric(merged_data['UNIT_SALE_PRICE'], errors='coerce')

# Filter de samengevoegde gegevens voor producten die voor meer dan 500 zijn verkocht en nooit zijn geretourneerd
gefilterde_data = merged_data[(merged_data['UNIT_SALE_PRICE'] > 500) & (merged_data['RETURN_QUANTITY'].isnull())]

# Selecteer de kolom voor productnaam en krijg unieke waarden
resultaat = gefilterde_data['PRODUCT_NAME'].unique()

# Toon het resulterende overzicht
pd.DataFrame(resultaat, columns=['PRODUCT_NAME'])


PRODUCT_NAME
0               Hailstorm Steel Irons
1            Hailstorm Titanium Irons
2          Lady Hailstorm Steel Irons
3       Lady Hailstorm Titanium Irons
4        Hailstorm Titanium Woods Set
5           Hailstorm Steel Woods Set
6   Lady Hailstorm Titanium Woods Set
7      Lady Hailstorm Steel Woods Set
8                           Star Dome
9                        Star Gazer 2
10                       Star Gazer 3
11                       Star Gazer 6
12                     Husky Rope 200

Geef een overzicht met daarin per (achternaam van) medewerker of hij/zij manager is of niet, door deze informatie toe te voegen als extra 'Ja/Nee'-kolom.<br>
Hint: gebruik een for-loop waarin je o.a. bepaalt of het woord 'Manager' in de functie (position_en) staat. (2 kolommen, 102 rijen).

In [34]:
# Maak een lege lijst om de resultaten op te slaan
resultaten = []

# Itereer over elke werknemer
for index, row in sales_staff.iterrows():
    # Controleer of het woord 'Manager' aanwezig is in de functie
    if 'Manager' in row['POSITION_EN']:
        resultaten.append('Ja')
    else:
        resultaten.append('Nee')

# Voeg de resultaten toe als een nieuwe kolom aan het sales_staff dataframe
sales_staff['Is Manager'] = resultaten

# Toon het resulterende overzicht
sales_staff[['LAST_NAME', 'Is Manager']]


LAST_NAME Is Manager
0    Savolainen        Nee
1           Kim        Nee
2          Choi        Nee
3           Kao        Nee
4          Fang        Nee
..          ...        ...
97     Iglesias        Nee
98        Lopes         Ja
99       Barros        Nee
100   Guimarães        Nee
101    Lindholm        Nee

[102 rows x 2 columns]

Met de onderstaande code laat je Python het huidige jaar uitrekenen.

In [35]:
from datetime import date
date.today().year

2024

Met de onderstaande code selecteer je op een bepaald jaartal uit een datum.

In [36]:
from datetime import datetime

date_str = '16-8-2013'
date_format = '%d-%m-%Y'
date_obj = datetime.strptime(date_str, date_format)

date_obj.year

2013

Geef met behulp van bovenstaande hulpcode een overzicht met daarin op basis van het aantal jaar dat iemand in dienst is of een medewerker ‘kort in dienst’ (minder dan 25 jaar in dienst) of een ‘lang in dienst’ (groter gelijk dan 12 jaar in dienst) is. Geef daarbij per medewerker in een aparte kolom zowel ‘kort in dienst’ als ‘lang in dienst’ aan. Gebruik (wederom) een for-loop.<br>
(2 kolommen, 102 rijen) 

In [37]:
from datetime import date

# Haal het huidige jaar op
current_year = date.today().year

# Zet de 'DATE_HIRED' kolom om naar datetime
sales_staff['DATE_HIRED'] = pd.to_datetime(sales_staff['DATE_HIRED'], dayfirst=True)

# Maak lege lijsten om de resultaten op te slaan
employees = []
tenure = []

# Itereer over elke werknemer
for index, row in sales_staff.iterrows():
    # Bereken het aantal jaren in dienst
    years_in_service = current_year - row['DATE_HIRED'].year
    
    # Categoriseer de diensttijd
    if years_in_service < 25:
        tenure_category = 'Korte diensttijd'
    else:
        tenure_category = 'Lange diensttijd'
    
    # Voeg de werknemer en diensttijd categorie toe aan de lijsten
    employees.append(row['LAST_NAME'])
    tenure.append(tenure_category)

# Maak een dataframe van de lijsten
resultaat = pd.DataFrame({'Werknemer': employees, 'Diensttijd Categorie': tenure})

# Toon het resulterende overzicht
resultaat


Employee Tenure Category
0    Savolainen    Short Tenure
1           Kim    Short Tenure
2          Choi    Short Tenure
3           Kao    Short Tenure
4          Fang    Short Tenure
..          ...             ...
97     Iglesias    Short Tenure
98        Lopes    Short Tenure
99       Barros    Short Tenure
100   Guimarães    Short Tenure
101    Lindholm    Short Tenure

[102 rows x 2 columns]